airflow/models/connection.py

```diff
@@ -377,3 +378,50 @@ def get_connection_from_secrets(cls, conn_id: str) -> 'Connection':
             if conn:
                 return conn
         raise AirflowNotFoundException(f"The conn_id `{conn_id}` isn't defined")
+
+    @classmethod
+    def get_connection_parameter_names(cls) -> Set[str]:
+        """Returns :class:`airflow.models.connection.Connection` constructor parameters."""
+        return {k for k in signature(cls.__init__).parameters.keys() if k != "self"}
+
+    @classmethod
+    def from_dict(cls, conn_id: str, conn_dict: Dict) -> 'Connection':
+        """
+        Create a connection from a dictionary.
+
+        :param conn_dict: dictionary representing a connection's attributes
+            e.g., {'conn_id': '', 'conn_type': '', 'login': '', ...}
+        :return: connection
+        """
+        if not isinstance(conn_dict, dict):
+            raise AirflowException(
+                f"Unexpected conn_dict type: {type(conn_dict)}. "
+                "The connection must be defined as a dictionary."
+            )
+
+        connection_parameter_names = cls.get_connection_parameter_names() | {"extra_dejson"}
+        current_keys = set(conn_dict.keys())
+        if not current_keys.issubset(connection_parameter_names):
+            illegal_keys = current_keys - connection_parameter_names
+            illegal_keys_list = ", ".join(sorted(illegal_keys))
+            raise AirflowException(
```

Comment:
Agreed with the same change on line 397, but this particular case shouldn't be a TypeError. It indicates that some keys in the dictionary aren't allowed. Maybe it can be a different type of exception though, what do you think?

Reply:
Oh, ValueError perhaps then?

Reply:
Sounds good 👍